src/man/fonehome.1
.\"  -*- nroff -*-
.\"
.Dd January 2, 2013
.Dt FONEHOME 1
.Os
.Sh NAME
.Nm fonehome
.Nd establish connection to fonehome server
.Sh SYNOPSIS
.Nm fonehome
.Bk -words
.Op Fl f Ar config-file
.Ek
.Nm fonehome
.Bk -words
.Fl I
.Ek
.Sh DESCRIPTION
.Nm
is a wrapper for the
.Xr ssh 1
program.
It creates a persistent no-pty
.Xr ssh 1
connection to one or more fonehome servers, with normal and reverse port forwarding
setup according to the configuration in
.Pa @fonehomeconf@ .
This allows private, bi-directional connections between the fonehome client and server,
while keeping those connections fully secured, without requiring any firewall setup.
.Pp
.Nm
is useful in situations where you have several machines
deployed in the field and want to maintain secure access to them from a central
operations server without any firewall hassles.
.Pp
Normally
.Nm
is run as a
.Xr systemd 1
service.
However, upon first installation (or whenever a server's host key changes),
a manual initialization step must be performed by running
.Nm
manually with the
.Fl I
flag.
This allows the administrator to confirm and accept the server host key(s).
.Sh OPTIONS
.Bl -tag -width Ds
.It Fl f Ar config-file
Use and alternate configuration file other than the default
.Pa @fonehomeconf@ .
.It Fl I
Perform first-time initialization when connecting to a new server.
.El
.Sh CONFIGURATION
The configuration file is a
.Xr bash 1
source file that defines the following shell variables:
.Bl -tag -width Ds
.It Pa SERVER
Defines the host to connect to.
This is the only variable that is strictly required.
.It Pa SSH_FLAGS
.Xr ssh 1
command line flags to use with the connection.
Typically this will contain reverse-forwarded ports to allow connections back from
the server to the client.
.It Pa USERNAME
SSH username.
Default is
.Ar @fonehomeuser@ .
.It Pa KEY_FILE
File containing the SSH private key.
This file must be unencrypted to allow for unattended operation; however,
it should be readable only by root.
Default is
.Ar @fonehomekey@ .
.It Pa RETRY_DELAY
How long (in seconds) to pause between connection attempts.
Default is
.Ar @fonehomeretry@
seconds.
.It Pa SYSLOG_TAG
Identifier to use when logging to
.Xr syslog 3 .
Default is
.Ar @fonehomename@ .
.It Pa SYSLOG_FACILITY
Facility to use when logging to
.Xr syslog 3 .
Default is
.Ar @fonehomelogfac@ .
.It Pa KNOWN_HOSTS_FILE
SSH known hosts file used to store recognized server public keys.
Default is
.Ar @fonehomehosts@ .
.El
.Sh MULTIPLE SERVERS
.Nm
suports connecting to multiple servers.
To enable multi-server support, simply declare
.Pa SERVER
as a
.Xr bash 1
array variable.
Then a separate connection will be maintained to each server in the array.
.Pp
When multiple servers are configured, the
.Pa USERNAME ,
.Pa SSH_FLAGS ,
.Pa KEY_FILE ,
and
.Pa RETRY_DELAY
variables may also be declared as arrays, in which case the corresponding
array elements will be used for each server.
If these variables are not arrays, or are shorter than the
.Pa SERVER
array, the last value will be used repeatedly.
Therefore, setting a simple (non-array) value results in that value being used for all servers.
.Sh FILES
.Bl -tag -width Ds -compact
.It Pa @fonehomeconf@
Default configuration file.
.It Pa @fonehomekey@
Default private SSH key file used to authenticate to the server.
.It Pa @fonehomehosts@
Default SSH known hosts file used to store and verify the server's public key.
.El
.Sh SEE ALSO
.Xr ssh 1 .
.Rs
.%T "fonehome: Remote access to machines behind firewalls"
.%O https://github.com/archiecobbs/fonehome/
.Re


CHANGES
Version Next

    - Added bash(1) completion scripts for fhssh(1) and fhscp(1)
    - Converted from /etc/init.d script to systemd unit file

Version 1.1.1 Released October 25, 2019

    - Filter zero bytes from /proc/X/cmdline to avoid bash warning
    - Run %post commands with reduced privileges to avoid security races

Version 1.1.0 Released January 7, 2016

    - Improvements to fhshow(1); support IPv6 and UDP sockets

Version 1.0.31 Released December 10, 2013

    - Add support for connecting to multiple servers (issue #1)
    - Log connection attempts and status to syslog (issue #2)

Version 1.0.28 Released February 18, 2013

    - Move home directory to /var/lib/fonehome-sever
    - Tighten security with no-pty and permitopen="0.0.0.0:9"
    - Remove sshd_config mods from %post script
    - Remove these flags no longer needed with newer SSH:
        -oUserKnownHostsFile=/dev/null
        -oGlobalKnownHostsFile=/dev/null
        -oNoHostAuthenticationForLocalhost=yes

Version 1.0.21 Released January 3, 2013

    - Move fonehome-server's homedir to /usr/share.
    - Fix several rpmlint warnings.
    - Fix bugs in man pages.

Version 1.0.15 Released January 2, 2013

    - Added man pages

Version 1.0.8 Released December 27, 2011

    - Initial release


src/scripts/fonehome.sh
#!/bin/bash

# Set constants and defaults
NAME='@fonehomename@'
USERNAME='@fonehomeuser@'
CONFIG_FILE='@fonehomeconf@'
KEY_FILE='@fonehomekey@'
RETRY_DELAY='@fonehomeretry@'
KNOWN_HOSTS_FILE='@fonehomehosts@'
FACILITY_PATTERN='^(auth|authpriv|cron|daemon|ftp|lpr|mail|news|security|user|uucp|local[0-7])$'
SYSLOG_TAG='@fonehomename@'
SYSLOG_FACILITY='@fonehomelogfac@'

# Usage message
usage()
{
    echo "Usage:" 1>&2
    echo "    ${NAME} [-I] [-f conf-file]" 1>&2
    echo "Options:" 1>&2
    echo "    -f    Specify alternate config file (default ${CONFIG_FILE})" 1>&2
    echo "    -I    Initialize: record server's public key (required on first connect)" 1>&2
}

# Log functions
perr()
{
    echo ${NAME}: ${1+"$@"} 1>&2
}

log()
{
    LEVEL="${1}"
    shift
    logger -t "${SYSLOG_TAG}" -p "${SYSLOG_FACILITY}"."${LEVEL}" ${1+"$@"}
    if [ -t 2 ]; then
        perr ${1+"$@"}
    fi
}

# Error function
errout()
{
    perr ${1+"$@"}
    if ! [ -t 2 ]; then
        log err ${1+"$@"}
    fi
    exit 1
}

# Bail on errors
set -e

# Enable job control
set -m

# Parse flags passed in on the command line
INITIALIZE="no"
while [ ${#} -gt 0 ]; do
    case "$1" in
        -f)
            shift
            CONFIG_FILE="$1"
            shift
            ;;
        -I)
            shift
            INITIALIZE="yes"
            ;;
        -h|--help)
            usage
            exit
            ;;
        --)
            shift
            break
            ;;
        -*)
            echo "${NAME}: unrecognized flag \`${1}'" 1>&2
            usage
            exit 1
            ;;
        *)
            break
            ;;
    esac
done
case "${#}" in
    0)
        ;;
    *)
        usage
        exit 1
        ;;
esac

# Read config file, which should set configuration variables
if ! [ -r "${CONFIG_FILE}" ]; then
    errout "can't read config file: ${CONFIG_FILE}"
fi
. "${CONFIG_FILE}"

# Verify required fields are set
if [[ -z SERVER ]]; then
    errout "no server(s) configured"
fi
if [[ -z USERNAME ]]; then
    errout "no username(s) configured"
fi
if [[ -z KEY_FILE ]]; then
    errout "no key file(s) configured"
fi

# Sanity check syslog facility name
if ! echo "${SYSLOG_FACILITY}" | grep -qE "${FACILITY_PATTERN}"; then
    errout "invalid syslog facility name \`${SYSLOG_FACILITY}'"
fi

# Make config variables into arrays
declare -a SERVER USERNAME KEY_FILE SSH_FLAGS RETRY_DELAY

# Get number of servers
declare -i NUM_SERVERS="${#SERVER[@]}"
declare -i i

# Extend all arrays, repeating last element as necessary
for ((i=0; $i < "${NUM_SERVERS}"; i++)); do
    if (( "${#USERNAME[@]}" <= $i )); then
        USERNAME[$i]="${USERNAME[$(($i - 1))]}"
    fi
    if (( "${#KEY_FILE[@]}" <= $i )); then
        KEY_FILE[$i]="${KEY_FILE[$(($i - 1))]}"
    fi
    if (( "${#SSH_FLAGS[@]}" <= $i )); then
        SSH_FLAGS[$i]="${SSH_FLAGS[$(($i - 1))]}"
    fi
    if (( "${#RETRY_DELAY[@]}" <= $i )); then
        RETRY_DELAY[$i]="${RETRY_DELAY[$(($i - 1))]}"
    fi
done

# Check required files
if [ -z "${KNOWN_HOSTS_FILE}" ]; then
    errout "no known hosts file configured"
fi
if [ "${INITIALIZE}" = 'yes' ]; then
    KNOWN_HOSTS_DIR=`dirname "${KNOWN_HOSTS_FILE}"`
    if ! [ -w "${KNOWN_HOSTS_DIR}" ]; then
        errout "known hosts file directory \`${KNOWN_HOSTS_DIR}' is unwritable"
    fi
else
    if ! [ -r "${KNOWN_HOSTS_FILE}" ]; then
        errout "known hosts file \`${KNOWN_HOSTS_FILE}' unreadable; did you run once with \`-I' flag?"
    fi
    if ! [ -s "${KNOWN_HOSTS_FILE}" ]; then
        errout "known hosts file \`${KNOWN_HOSTS_FILE}' is empty; did you run once with \`-I' flag for each server?"
    fi
fi
for ((i=0; $i < "${NUM_SERVERS}"; i++)); do
    if ! [ -r "${KEY_FILE[$i]}" ]; then
        errout "key file \`${KEY_FILE[$i]}' for server ${SERVER[$i]} is unreadable"
    fi
done

# Doing initialization?
if [ "${INITIALIZE}" = "yes" ]; then
    for ((i=0; $i < "${NUM_SERVERS}"; i++)); do
        perr "confirming server key for ${SERVER[$i]}"
        ssh -24xaT -oStrictHostKeyChecking=ask -oUserKnownHostsFile="${KNOWN_HOSTS_FILE}" \
          -oVisualHostKey=yes -oCheckHostIP=no -i "${KEY_FILE[$i]}" ${SSH_FLAGS[$i]} "${USERNAME[$i]}"@"${SERVER[$i]}"
    done
    exit 0
fi

# Subshell - there is one of these for each server
subshell()
{
    # Don't die on error
    set +e

    # Which server am I?
    INDEX="${1}"

    # Create temporary file to hold output of ssh(1) command
    OUTPUT_FILE=`mktemp -q /tmp/fonehome.XXXXXXXX`
    if [ $? -ne 0 ]; then
        log error "can't create temporary file"
        OUTPUT_FILE="/dev/null"
    fi

    # Clean up output file on exit
    trap "if [ \"${OUTPUT_FILE}\" != '/dev/null' ]; then rm -f \"${OUTPUT_FILE}\"; fi; exit" INT TERM

    # Connect to server; reconnect if we get disconnected
    while true; do
        log info initiating connection to "${SERVER[$INDEX]}"
        ssh -24xaTnN -oBatchMode=yes -oExitOnForwardFailure=yes \
          -oStrictHostKeyChecking=yes -oUserKnownHostsFile="${KNOWN_HOSTS_FILE}" \
          -oServerAliveInterval=60 -oServerAliveCountMax=5 -oTCPKeepAlive=yes \
          -oCheckHostIP=no -i "${KEY_FILE[$INDEX]}" ${SSH_FLAGS[$INDEX]} "${USERNAME[$INDEX]}"@"${SERVER[$INDEX]}" >"${OUTPUT_FILE}" 2>&1
        if [ -s "${OUTPUT_FILE}" ]; then
            log warn "connection to ${SERVER[$INDEX]} failed: `cat \"${OUTPUT_FILE}\"`"
        else
            log warn "connection to ${SERVER[$INDEX]} failed"
        fi
        sleep "${RETRY_DELAY[$INDEX]}"
    done
}

# This function is used to ensure the subshells are killed when this script is killed
killshells()
{
    jobs -p | sed 's/^/-/g' | xargs -r kill --
    log info shutting down
    wait
    exit
}

# Clean up subshells on exit
trap killshells INT TERM

# Start the subshells
for ((i=0; $i < "${NUM_SERVERS}"; i++)); do
    subshell $i &
done

# Now we just hang out
wait

